File: heuristic_func.py
import numpy as np
from multiprocessing import Pool

##############################################################################
# Heuristic Function###########################################################


def heuristic_m(node, goal, D=1):  # Manhattan Distance
    dx = abs(node.position[0] - goal.position[0])
    dy = abs(node.position[1] - goal.position[1])
    return D * (dx + dy)


def heuristic_d(node, goal, D=1, D2=2 ** 0.5):  # Diagonal Distance
    dx = abs(node.position[0] - goal.position[0])
    dy = abs(node.position[1] - goal.position[1])
    return D * (dx + dy) + (D2 - 2 * D) * min(dx, dy)


def heuristic_e(node, goal):  # Euclidean Distance
    dx = abs(node.position[0] - goal.position[0])
    dy = abs(node.position[1] - goal.position[1])
    return np.sqrt(dx ** 2 + dy ** 2)


def get_around_index(node, r, maze):
    h = len(maze)
    w = len(maze[0])
    idx = []
    for i in range(node.position[0] - r, node.position[0] + r + 1):
        for j in range(node.position[1] - r, node.position[1] + r + 1):
            radius = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
            if 0 <= i < h and 0 <= j < w and r >= radius:
                idx.append([i, j])
    return idx


def get_index_to_goal(node, goal):
    idx = []
    if goal.position[0] - node.position[0] == goal.position[1] - node.position[1]:
        idx = np.vstack((np.arange(node.position[0], goal.position[0] + 1, 1),
                         np.arange(node.position[1], goal.position[1] + 1, 1),
                         np.sqrt(2) * (np.arange(node.position[0], goal.position[0] + 1, 1)))).astype(np.int64).T.tolist()
        return [idx]
    else:
        upper_idx = []
        lower_idx = []
        # 목적지가 1사분면에 있는 경우
        if node.position[0] <= goal.position[0] and node.position[1] <= goal.position[1]:
            for i in range(node.position[0], goal.position[0] + 1, 1):
                for j in range(node.position[1], goal.position[1] + 1, 1):
                    if (goal.position[1] - node.position[1] + 1) * (i - goal.position[0]) >= \
                            (goal.position[0] - node.position[0] + 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        upper_idx.append([i, j, r])
                    if (goal.position[1] - node.position[1] + 1) * (i - goal.position[0]) <= \
                            (goal.position[0] - node.position[0] + 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        lower_idx.append([i, j, r])
        # 목적지가 2사분면에 있는 경우
        elif node.position[0] >= goal.position[0] and node.position[1] <= goal.position[1]:
            for i in range(node.position[0], goal.position[0] - 1, -1):
                for j in range(node.position[1], goal.position[1] + 1, 1):
                    if (goal.position[1] - node.position[1] + 1) * (i - goal.position[0]) >= \
                            (goal.position[0] - node.position[0] - 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        upper_idx.append([i, j, r])
                    if (goal.position[1] - node.position[1] + 1) * (i - goal.position[0]) <= \
                            (goal.position[0] - node.position[0] - 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        lower_idx.append([i, j, r])
        # 목적지가 3사분면에 있는 경우
        elif node.position[0] >= goal.position[0] and node.position[1] >= goal.position[1]:
            for i in range(node.position[0], goal.position[0] - 1, -1):
                for j in range(node.position[1], goal.position[1] - 1, -1):
                    if (goal.position[1] - node.position[1] - 1) * (i - goal.position[0]) >= \
                            (goal.position[0] - node.position[0] - 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        upper_idx.append([i, j, r])
                    if (goal.position[1] - node.position[1] - 1) * (i - goal.position[0]) <= \
                            (goal.position[0] - node.position[0] - 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        lower_idx.append([i, j, r])
        # 목적지가 4사분면에 있는 경우
        elif node.position[0] <= goal.position[0] and node.position[1] >= goal.position[1]:
            for i in range(node.position[0], goal.position[0] + 1, 1):
                for j in range(node.position[1], goal.position[1] - 1, -1):
                    if (goal.position[1] - node.position[1] - 1) * (i - goal.position[0]) >= \
                            (goal.position[0] + node.position[0] + 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        upper_idx.append([i, j, r])
                    if (goal.position[1] - node.position[1] - 1) * (i - goal.position[0]) <= \
                            (goal.position[0] + node.position[0] + 1) * (j - goal.position[1]):
                        r = np.sqrt((i - node.position[0]) ** 2 + (j - node.position[1]) ** 2)
                        lower_idx.append([i, j, r])
        return [upper_idx, lower_idx]


def heuristic_around_obstacle(node, goal, origin_maze, potential_map):
    height = len(origin_maze)
    width = len(origin_maze[0])
    map_size = np.sqrt(height * width)
    h = heuristic_e(node, goal)
    idx = get_around_index(node, int(map_size / 20) + 1, origin_maze)
    h1 = 0
    for i in range(len(idx)):
        if origin_maze[idx[i][0]][idx[i][1]] == 1:
            h1 += 1.5 * potential_map[idx[i][0]][idx[i][1]]
        elif origin_maze[idx[i][0]][idx[i][1]] == 2:
            h1 += 4 * potential_map[idx[i][0]][idx[i][1]]
        elif origin_maze[idx[i][0]][idx[i][1]] == 3:
            h1 += potential_map[idx[i][0]][idx[i][1]]
    if len(idx) != 0:
        h = h1 / len(idx) * np.sqrt(h)
    return h


def heuristic(node, goal, origin_maze, potential_map):
    # print(len(origin_maze))
    # print(len(origin_maze[0]))
    h = heuristic_e(node, goal)
    idx = get_index_to_goal(node, goal)
    R = heuristic_e(node, goal)
    beta = 40  # 가까이 있는 장애물 보정계수
    # 대각선이 존재하는 경우
    if len(idx) == 1:
        h1 = 0
        for i in range(len(idx[0])):
            h1 += (1 - (min(idx[0][i][2], R) / (R + 0.01)) ** 3) * potential_map[idx[0][i][0]][[idx[0][i][1]]][0]
        if len(idx[0]) == 0:
            return 1
        # print("h11:", h1)
        h *= h1 / len(idx[0]) * np.sqrt(R)        # 경로상의 node 수 보정
    else:
        h1 = 0
        h2 = 0
        for i in range(len(idx[0])):
            h1 += (1 - (min(idx[0][i][2], R) / (R + 0.01)) ** 3) * potential_map[idx[0][i][0]][[idx[0][i][1]]][0]
        for i in range(len(idx[1])):
            h2 += (1 - (min(idx[1][i][2], R) / (R + 0.01)) ** 3) * potential_map[idx[1][i][0]][[idx[1][i][1]]][0]
        if len(idx[0]) != 0 and len(idx[1]) != 0:
            h1 = (h1 / len(idx[0])) * np.sqrt(h)        # 경로상의 node 수 보정
            h2 = (h2 / len(idx[1])) * np.sqrt(h)        # 경로상의 node 수 보정
            # print("h1:", h1, "\nh2:", h2)
        h *= min(h1, h2)
    if type(h) is not int:
        h = int(h.astype(np.int64))
    if h <= 0:
        h = 1

    h += beta * heuristic_around_obstacle(node, goal, origin_maze, potential_map)
    return h

File: C++/Heuristic_func.cpp
#include "Heuristic_func.h"
#include <tuple>
#include <cmath>

double heuristic_m(p node, p goal, double d = 1)
{
	int dx = abs(node.first - goal.first);
	int dy = abs(node.second - goal.second);
	return d * (dx + dy);
}

double heuristic_d(p node, p goal, double d1 = 1, double d2 = sqrt(2))
{
	int dx = abs(node.first - goal.first);
	int dy = abs(node.second - goal.second);
	return d1 * (dx + dy) + (d2 - 2 * d1) * min(dx, dy);
}

double heuristic_e(p node, p goal)
{
	int dx = abs(node.first - goal.first);
	int dy = abs(node.second - goal.second);
	return sqrt(dx * dx + dy * dy);
}

vector<p> get_around_index(p node, double r, Map map)
{
	pair<ll, ll> map_size = { map.size() , map[0].size() };
	vector<p> idx;
	for (ll i = map_size.first - r; i < map_size.first + 1; i++)
	{
		for (ll j = map_size.second - r; j < map_size.second + 1; j++)
		{
			double dist = sqrt(pow(i - map_size.first, 2) + pow(j - map_size.second, 2));
			if ((0 <= i && i < map_size.first) && (0 <= j && j < map_size.second) && (r >= dist))
			{
				idx.push_back({ i, j });
			}
		}
	}
	return idx;
}

vector<MapIdx> get_index_to_goal_sq(p node, p goal)
{
	vector<MapIdx> idx;
	for (int i = 0; i < goal.first - node.first; i++)
		idx.push_back({ { i + node.first, i + node.second }, sqrt(2) * (i + node.first) });
	return idx;
}

pair<vector<MapIdx>, vector<MapIdx>> get_index_to_goal_rect(p node, p goal)
{
	vector<pair<p, double>> upperIdx;
	vector<pair<p, double>> lowerIdx;
	if (node.first <= goal.first and node.second <= goal.second)
	{
		for (int i = node.first; i < goal.first + 1; i++)
		{
			for (int j = node.second; j < goal.second + 1; j++)
			{
				if ((goal.second - node.second + 1) * (i - goal.first) >= (goal.first - node.first + 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					upperIdx.push_back({ {i, j}, r });
				}
				if ((goal.second - node.second + 1) * (i - goal.first) <= (goal.first - node.first + 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					lowerIdx.push_back({ {i, j}, r });
				}
			}
		}
	}
	else if (node.first >= goal.first and node.second <= goal.second)
	{
		for (int i = node.first; i > goal.first - 1; i--)
		{
			for (int j = node.second; j < goal.second + 1; j++)
			{
				if ((goal.second - node.second + 1) * (i - goal.first) >= (goal.first - node.first - 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					upperIdx.push_back({ {i, j}, r });
				}
				if ((goal.second - node.second + 1) * (i - goal.first) <= (goal.first - node.first - 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					lowerIdx.push_back({ {i, j}, r });
				}
			}
		}
	}
	else if (node.first >= goal.first and node.second >= goal.second)
	{
		for (int i = node.first; i > goal.first - 1; i--)
		{
			for (int j = node.second; j > goal.second - 1; j--)
			{
				if ((goal.second - node.second - 1) * (i - goal.first) >= (goal.first - node.first - 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					upperIdx.push_back({ {i, j}, r });
				}
				if ((goal.second - node.second - 1) * (i - goal.first) <= (goal.first - node.first - 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					lowerIdx.push_back({ {i, j}, r });
				}
			}
		}
	}
	else if (node.first <= goal.first and node.second >= goal.second)
	{
		for (int i = node.first; i < goal.first + 1; i++)
		{
			for (int j = node.second; j > goal.second - 1; j--)
			{
				if ((goal.second - node.second - 1) * (i - goal.first) >= (goal.first - node.first + 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					upperIdx.push_back({ {i, j}, r });
				}
				if ((goal.second - node.second - 1) * (i - goal.first) <= (goal.first - node.first + 1) * (j - goal.second))
				{
					double r = sqrt(pow(i - node.first, 2) + pow(j - node.second, 2));
					lowerIdx.push_back({ {i, j}, r });
				}
			}
		}
	}
	pair<vector<MapIdx>, vector<MapIdx>> idx = { upperIdx, lowerIdx };
	return idx;
}

double heuristic_around_obstacle(p node, p goal, Map map, Map potential_map)
{
	pair<ll, ll> map_size = { map.size() , map[0].size() };
	double r = 3;
	double R = heuristic_e(node, goal);
	vector<p> idx = get_around_index(node, r, map);
	double h = 0;
	for (int i = 0; i < idx.size(); i++)
	{
		switch (map[idx[i].first][idx[i].second])
		{
		case 1:
			h += 1.5 * potential_map[idx[i].first][idx[i].second];
			break;
		case 2:
			h += 4 * potential_map[idx[i].first][idx[i].second];
			break;
		case 3:
			h += 1 * potential_map[idx[i].first][idx[i].second];
			break;
		default:
			break;
		}
	}
	if (idx.size() != 0) {
		h = (h / idx.size()) * sqrt(R);	
	}
	return h;
}

double heuristic(p node, p goal, Map map, Map potential_map)
{
	double h = heuristic_e(node, goal);
	double R = heuristic_e(node, goal);
	double beta = 40;

	if (goal.first - node.first == goal.second - node.second)
	{
		vector<MapIdx> idx = get_index_to_goal_sq(node, goal);
		if (idx.size() == 0)
			return 1;
		double h_add = 0;
		for (int i = 0; i < idx.size(); i++)
		{
			h_add += (1 - pow(min(idx[i].second, R) / (R + 0.01), 3)) * potential_map[idx[i].first.first][idx[i].first.second];
		}
		h *= (h_add / idx.size()) * sqrt(R);
	}
	else
	{
		pair<vector<MapIdx>, vector<MapIdx>> idx = get_index_to_goal_rect(node, goal);
		if (idx.first.size() == 0 || idx.second.size() == 0)
			return 1;
		double h_add1 = 0, h_add2 = 0;
		for (int i = 0; i < idx.first.size(); i++)
		{
			h_add1 += (1 - pow(min(idx.first[i].second, R) / (R + 0.01), 3)) * potential_map[idx.first[i].first.first][idx.first[i].first.second];
		}
		for (int i = 0; i < idx.second.size(); i++)
		{
			h_add2 += (1 - pow(min(idx.second[i].second, R) / (R + 0.01), 3)) * potential_map[idx.second[i].first.first][idx.second[i].first.second];
		}
		h_add1 = (h_add1 / idx.first.size()) * sqrt(R);
		h_add2 = (h_add2 / idx.second.size()) * sqrt(R);
	}

	h += heuristic_around_obstacle(node, goal, map, potential_map);
	return h;
}